geoips/pydantic/products.py

```diff
@@ -0,0 +1,231 @@
+
+    @model_validator(mode="before")
+    def _validate_plugin_name(cls, values: Dict[str, Any]) -> Dict[str, Any]:
+        """Validate plugin name."""
+        if not values:
+            raise ValueError("The plugin_name cannot be empty")
+        plugin_name = values.get("name", "").lower()
+        plugin_type = values.get("type", " ").lower()
+
+        valid_plugin_names = get_plugin_names(plugin_type)
+        if plugin_name not in valid_plugin_names:
+            raise ValueError(
+                f"{plugin_name} is invalid. \n\t"
+                f"Must be one of {sorted(valid_plugin_names)}"
+            )
+
+        return values
+
+    @model_validator(mode="before")
+    def _validate_plugin_arguments(cls, values: Dict[str, Any]) -> Dict[str, Any]:
+        """
+        Validate & organize details for each step   .
+
+        Parameters
+        ----------
+        values : dict
+            A dictionary of plugin data. The key is plugin type, and
+            the value consists of plugin name & arguments
+
+        Returns
+        -------
+        values : dict
+            A validated and structured dictionary with the following fields:
+
+            - `type` : str
+                The type of the plugin.
+            - `name` : str
+                The name of the plugin.
+            - `arguments` : dict
+                The arguments associated with the plugin.
+
+        """
+        if not values:
+            raise ValueError("Empty : Missing step details")
+
+        # arguments = values.get("arguments", {})
+        # Delegate arguments validation to each plugin type argument class
+        plugin_type = values.get("type", " ").lower()
+        plugin_type_camel_case = "".join(
+            [word.capitalize() for word in plugin_type.split("_")]
+        )
+        plugin_arguments_model_name = f"{plugin_type_camel_case}ArgumentsModel"
+        plugin_arguments_model = globals().get(plugin_arguments_model_name)
+        if plugin_arguments_model is None:
+            raise ValueError(
+                f"\n\n\tThe plugin type argument class"
+                f"{plugin_arguments_model_name} is not defined\n\n"
+            )
+        plugin_arguments_model(**values.get("arguments", {}))
+
+        return values
+
+
+class ProductStepModel(BaseModel):
+    """Validate and process a sequence of steps with their data."""
+
+    definition: ProductStepDefinitionModel = Field(..., description="Sequence of steps")
+
+    @model_validator(mode="before")
+    def _plugin_name_validator(cls, values: Dict[str, Any]) -> Dict[str, Any]:
+        """
+        Validate user input for the plugin type.
+
+        Parameters
+        ----------
+        values : dict
+            The step data containing step name (plugin type) and its content.
+
+        Returns
+        -------
+        dict
+            The processed step data after step name validation
```

Comment:
Info about the format of the returned dict could be helpful since it is a standard format